File: evals/swebench/prices.ts
/**
 * The committed price table the scorecard grades cost with.
 *
 * Prices are USD per million tokens, as published for the OpenAI API. They are
 * committed rather than fetched so that re-scoring an old run reproduces the
 * old number: a price change must be an edit here, with a date, not a silent
 * drift in someone's report.
 *
 * Verified 2026-08-19 against:
 *   - https://openai.com/index/advancing-the-price-performance-frontier-with-gpt-5-6/
 *     (the GPT-5.6 launch post: Sol at $5 / $30 per 1M input / output)
 *   - https://www.orcarouter.ai/blog/gpt-5-6-sol-pricing (verified 2026-08-18:
 *     $5.00 input on a cache miss, $0.50 cached input, $30.00 output)
 *   - https://openrouter.ai/openai/gpt-5.6-sol (quotes the same base rates
 *     under a promotional 50% discount, cache read at half of $0.50)
 *
 * A seat the table does not name has undefined cost with an explicit
 * `unpriced` note, never silently at another model's rate.
 *
 * Jev is priced from the journal beside the seat, never folded into the
 * seat's number: `claim-demanded` and `supervisor-settled` carry the usage of
 * the two harness classifiers, and a `cell-call-settled` of the `jev` flow
 * carries the usage of an agent's own call. `decision-settled` carries the
 * same readings without usage and is never priced.
 *
 * @since 0.1.0
 */

/**
 * One model's USD price per million tokens.
 *
 * @category models
 * @since 0.1.0
 */
export interface Price {
  /** Uncached input tokens, USD per 1M. */
  readonly input: number
  /** Cache-read input tokens, USD per 1M. */
  readonly cachedInput: number
  /** Output tokens, including reasoning tokens, USD per 1M. */
  readonly output: number
  /** Where the numbers came from, and when they were checked. */
  readonly source: string
}

/**
 * The gateway id of Jev, as `Evaluator.defaultModel` names it. Every Jev
 * reading a journal records is priced under this row, whichever seat the
 * run's model turns were on.
 *
 * @category constants
 * @since 0.1.0
 */
export const jevModel = "typesafe-ai/jev"

/**
 * The table, keyed by both the bare model id and the seat spelling the flows
 * CLI uses, because a run's journal records the seat and a codex run records
 * the model.
 *
 * @category constants
 * @since 0.1.0
 */
export const prices: Record<string, Price> = {
  "gpt-5.6-sol": {
    input: 5,
    cachedInput: 0.5,
    output: 30,
    source: "OpenAI API list price, verified 2026-08-19"
  },
  "openai:gpt-5.6-sol": {
    input: 5,
    cachedInput: 0.5,
    output: 30,
    source: "OpenAI API list price, verified 2026-08-19"
  },
  // The standard-tier short-context rate; prompts over 272K input tokens are
  // billed at 2x input and 1.5x output for the whole request, which no row here
  // models. A run served by the ChatGPT subscription (`SMITHERS_OPENAI_AUTH=
  // chatgpt`) is billed by the month, not by the token: the price below is
  // what the same tokens WOULD cost on the API, and `evals/harbor` leaves such
  // runs unpriced rather than quote it as spend.
  "gpt-6-sol": {
    input: 2,
    cachedInput: 0.2,
    output: 10,
    source: "OpenAI API list price, https://developers.openai.com/api/docs/models/gpt-6-sol, verified 2026-09-22"
  },
  "openai:gpt-6-sol": {
    input: 2,
    cachedInput: 0.2,
    output: 10,
    source: "OpenAI API list price, https://developers.openai.com/api/docs/models/gpt-6-sol, verified 2026-09-22"
  },
  // Jev, the decision model the harness asks through the Vercel AI Gateway:
  // the completion brake, the per-frame supervisor and the agent-callable
  // `jev` flow all bill against it. The gateway quotes per-token pricing of
  // 0.000000042 USD input and 0 output (that is 0.042 USD per 1M input; the
  // gateway's model page rounds it to $0.04/1M). Jev has no cached-input
  // tier, so the cache-read rate is the input rate.
  [jevModel]: {
    input: 0.042,
    cachedInput: 0.042,
    output: 0,
    source:
      "Vercel AI Gateway list price, https://ai-gateway.vercel.sh/v1/models/typesafe-ai/jev (pricing.input 0.000000042 per token, output 0) and https://vercel.com/ai-gateway/models ($0.04/1M input, $0.00/1M output), verified 2026-09-22"
  }
}

/**
 * Computes USD for one run's token counts.
 *
 * `inputTokens` is the provider's total prompt count and already contains
 * `cachedInputTokens`, so the cached share is billed at the cache-read rate and
 * only the remainder at the input rate. Reasoning tokens are part of the output
 * count and are not billed twice.
 *
 * @category constructors
 * @since 0.1.0
 */
export const usd = (
  model: string | undefined,
  tokens: { readonly inputTokens: number; readonly cachedInputTokens: number; readonly outputTokens: number }
): { readonly usd: number | undefined; readonly source: string } => {
  const price = model === undefined ? undefined : prices[model]
  if (price === undefined) {
    return { usd: undefined, source: `unpriced: no committed price for ${model ?? "an unrecorded model"}` }
  }
  const uncached = Math.max(0, tokens.inputTokens - tokens.cachedInputTokens)
  const total = (uncached * price.input + tokens.cachedInputTokens * price.cachedInput
    + tokens.outputTokens * price.output) / 1_000_000
  return { usd: Math.round(total * 10_000) / 10_000, source: price.source }
}

File: packages/backend/modelprice/prices.go
// Package modelprice is the single checked-in price table for every model
// Smithers runs on platform keys. Every metered model call computes its cost
// here; nothing else in the tree holds a price.
//
// Prices are list prices per million tokens in micro-dollars (1 USD =
// 1_000_000). A model missing from the table, or whose price for long
// prompts is not declared, is refused by the metering point, never
// estimated: an unpriced model is an unmetered bill.
package modelprice

import (
	"errors"
	"math/big"
	"strings"
)

// Rates is one rate card. Every class a provider can report is priced; a
// provider without a cheaper cache class is entered at its input rate.
type Rates struct {
	// InputPerMTok / OutputPerMTok are micro-dollars per million tokens.
	InputPerMTok  int64
	OutputPerMTok int64
	// CacheReadPerMTok / CacheWritePerMTok apply to prompt-cache tokens the
	// provider reports separately from InputPerMTok.
	CacheReadPerMTok  int64
	CacheWritePerMTok int64
}

// ContextPricing declares how a model's price depends on prompt size. The
// zero value is unknown, and a model with unknown context pricing is refused.
type ContextPricing int

const (
	contextUnknown ContextPricing = iota
	// ContextFlat is one rate card at every prompt size the model accepts.
	ContextFlat
	// ContextTiered charges the whole call at LongContext once the prompt
	// reaches LongContextFrom input tokens.
	ContextTiered
)

// Price is one model's rate card.
type Price struct {
	// Provider is the upstream the model is served by (anthropic, openai,
	// cerebras, openrouter, vercel).
	Provider string
	Rates
	Context ContextPricing
	// LongContextFrom is the prompt size, in input tokens (uncached, cache
	// read and cache write together), from which LongContext prices every
	// token of the call. Only for ContextTiered.
	LongContextFrom int64
	LongContext     Rates
	// FlatPerCall is charged once per call for endpoints that report no
	// token usage (the Jev evaluation model).
	FlatPerCall int64
}

// Usage is what a provider reported for one call.
type Usage struct {
	InputTokens      int64
	OutputTokens     int64
	CacheReadTokens  int64
	CacheWriteTokens int64
}

// PromptTokens is every input token of the call, whatever its cache class:
// the size providers compare with a long-context threshold.
func (u Usage) PromptTokens() int64 {
	return u.InputTokens + u.CacheReadTokens + u.CacheWriteTokens
}

// usd converts a dollar amount to micro-dollars: usd(2.5) == 2_500_000.
func usd(dollars float64) int64 { return int64(dollars*1_000_000 + 0.5) }

// OpenAILongContextFrom is where OpenAI's long-context rates start. OpenAI
// states ">272K input tokens"; OpenRouter lists the same rates from
// min_prompt_tokens 272000. The premium is charged from 272,000 so the
// boundary token itself can never be under-charged.
const OpenAILongContextFrom = 272_000

// Table is keyed by the model id the caller sends on the wire.
//
// Sources, read 2026-09-25:
//   - Anthropic: https://platform.claude.com/docs/en/about-claude/pricing
//     ("Long context pricing": Claude 4.6 and later include the 1M context
//     window at standard pricing) and
//     https://platform.claude.com/docs/en/build-with-claude/context-windows
//     (Sonnet 4.5 and Haiku 4.5 have a 200k window; the proxy refuses the
//     context-1m beta, and OpenRouter's 1M Sonnet 4.5 is tiered). Cache write 1.25x input (5-minute; the 1-hour cache
//     is refused), cache read 0.1x (0.05x Opus 5.5, 0.025x Fable 5.1).
//   - OpenAI: https://developers.openai.com/api/docs/pricing and each
//     https://developers.openai.com/api/docs/models/<id> page: prompts over
//     272K input tokens are priced at 2x input and cache rates and 1.5x
//     output for the full request. GPT-5.6 and later bill cache writes at
//     1.25x input; earlier models add no cache-write charge.
//   - Cerebras: https://www.cerebras.ai/pricing (developer tier; context
//     131k/128k per https://inference-docs.cerebras.ai/models/overview).
//     No cache price is published, so cached tokens are charged as input.
//   - OpenRouter: https://openrouter.ai/api/v1/models.
var Table = map[string]Price{
	"claude-fable-5-1":  anthropic(10, 50, 0.025),
	"claude-fable-5":    anthropic(10, 50, 0.1),
	"claude-opus-5-5":   anthropic(4, 20, 0.05),
	"claude-opus-5":     anthropic(5, 25, 0.1),
	"claude-opus-4-8":   anthropic(5, 25, 0.1),
	"claude-opus-4-7":   anthropic(5, 25, 0.1),
	"claude-opus-4-6":   anthropic(5, 25, 0.1),
	"claude-sonnet-5":   anthropic(2, 10, 0.1),
	"claude-sonnet-4-6": anthropic(3, 15, 0.1),
	// 200k context window on Anthropic, where the 1M beta is refused; 1M on
	// OpenRouter, which charges 2x input and cache and 1.5x output from
	// 200,000 prompt tokens (https://openrouter.ai/api/v1/models).
	"claude-sonnet-4-5": anthropicTiered(3, 15, 200_000),
	"claude-haiku-4-5":  anthropic(1, 5, 0.1),

	"gpt-6-astra": openaiTiered(10, 1, 12.5, 50),
	"gpt-6-sol":   openaiTiered(2, 0.2, 2.5, 10),
	"gpt-6-luna":  openaiTiered(0.1, 0.01, 0.125, 0.5),
	// Promotional through at least 2026-11-21; re-check before then.
	"gpt-5.6-sol":   openaiTiered(4, 0.4, 5, 20),
	"gpt-5.6-terra": openaiTiered(2, 0.2, 2.5, 12),
	"gpt-5.6-luna":  openaiTiered(0.2, 0.02, 0.25, 1.2),
	"gpt-5.5":       openaiTiered(5, 0.5, 5, 30),
	// 128k context window, one rate card.
	"gpt-4o": {Provider: "openai", Context: ContextFlat, Rates: Rates{InputPerMTok: usd(2.5), OutputPerMTok: usd(10), CacheReadPerMTok: usd(1.25), CacheWritePerMTok: usd(2.5)}},

	"gpt-oss-120b": flat("cerebras", 0.35, 0.75),
	"qwen-3.8-27b": flat("cerebras", 0.99, 1.49),
	// OpenRouter lists 0.15/0.60 (cache read 0.075); the ceiling sent with
	// the call is this row, so no routed provider costs more.
	"gpt-oss-120b@openrouter": flat("openrouter", 0.35, 0.75),

	// Vercel AI Gateway evaluation model: no token usage on the wire, flat
	// per call (VERIFY against the gateway invoice).
	"typesafe-ai/jev": {Provider: "vercel", Context: ContextFlat, FlatPerCall: usd(0.002)},
}

func anthropic(in, out, cacheRead float64) Price {
	return Price{Provider: "anthropic", Context: ContextFlat, Rates: Rates{
		InputPerMTok:      usd(in),
		OutputPerMTok:     usd(out),
		CacheReadPerMTok:  usd(in * cacheRead),
		CacheWritePerMTok: usd(in * 1.25),
	}}
}

// anthropicTiered adds Anthropic's former long-context premium: 2x input and
// cache rates and 1.5x output for the whole call from the threshold.
func anthropicTiered(in, out float64, from int64) Price {
	price := anthropic(in, out, 0.1)
	price.Context, price.LongContextFrom = ContextTiered, from
	price.LongContext = Rates{InputPerMTok: usd(in * 2), OutputPerMTok: usd(out * 1.5), CacheReadPerMTok: usd(in * 0.2), CacheWritePerMTok: usd(in * 2.5)}
	return price
}

// openaiTiered is an OpenAI rate card with the long-context premium: 2x
// input and cache rates and 1.5x output from OpenAILongContextFrom.
func openaiTiered(in, cacheRead, cacheWrite, out float64) Price {
	return Price{
		Provider:        "openai",
		Context:         ContextTiered,
		Rates:           Rates{InputPerMTok: usd(in), OutputPerMTok: usd(out), CacheReadPerMTok: usd(cacheRead), CacheWritePerMTok: usd(cacheWrite)},
		LongContextFrom: OpenAILongContextFrom,
		LongContext:     Rates{InputPerMTok: usd(in * 2), OutputPerMTok: usd(out * 1.5), CacheReadPerMTok: usd(cacheRead * 2), CacheWritePerMTok: usd(cacheWrite * 2)},
	}
}

// flat is one rate card with no published cache discount.
func flat(provider string, in, out float64) Price {
	return Price{Provider: provider, Context: ContextFlat, Rates: Rates{InputPerMTok: usd(in), OutputPerMTok: usd(out), CacheReadPerMTok: usd(in), CacheWritePerMTok: usd(in)}}
}

// Lookup returns the price for a model id. Provider-prefixed ids such as
// "anthropic/claude-sonnet-5" (OpenRouter and the Vercel gateway) resolve to
// the bare id. A model whose context pricing is unknown is not found.
func Lookup(model string) (Price, bool) {
	model = strings.TrimSpace(model)
	price, ok := Table[model]
	if !ok {
		if i := strings.LastIndex(model, "/"); i >= 0 {
			price, ok = Table[model[i+1:]]
		}
	}
	if !ok || !price.known() {
		return Price{}, false
	}
	return price, true
}

func (p Price) known() bool {
	switch p.Context {
	case ContextFlat:
		return true
	case ContextTiered:
		return p.LongContextFrom > 0
	}
	return false
}

// RatesFor is the rate card that prices a call with this many prompt tokens.
func (p Price) RatesFor(promptTokens int64) Rates {
	if p.Context == ContextTiered && promptTokens >= p.LongContextFrom {
		return p.LongContext
	}
	return p.Rates
}

// Maximum is the dearest usage a call can report when its prompt is at most
// promptTokens and its output at most outputTokens: every prompt token is
// placed in the dearest input class of the rate card that prompt size
// selects. Any split the provider reports costs no more than this, provided
// no long-context rate is below its standard rate (checked by the tests).
func (p Price) Maximum(promptTokens, outputTokens int64) Usage {
	rates := p.RatesFor(promptTokens)
	usage := Usage{OutputTokens: outputTokens}
	switch max(rates.InputPerMTok, rates.CacheReadPerMTok, rates.CacheWritePerMTok) {
	case rates.CacheWritePerMTok:
		usage.CacheWriteTokens = promptTokens
	case rates.CacheReadPerMTok:
		usage.CacheReadTokens = promptTokens
	default:
		usage.InputTokens = promptTokens
	}
	return usage
}

// CostNanos computes a provider charge in integer USD nanos. The prompt size
// selects the rate card; the sum is rounded upward once, preserving sub-cent
// value without floats.
func CostNanos(price Price, usage Usage) (int64, error) {
	if usage.InputTokens < 0 || usage.OutputTokens < 0 || usage.CacheReadTokens < 0 || usage.CacheWriteTokens < 0 {
		return 0, errors.New("negative model usage")
	}
	if !price.known() {
		return 0, errors.New("model context pricing is unknown")
	}
	if price.FlatPerCall < 0 {
		return 0, errors.New("negative model price")
	}
	prompt := new(big.Int).Add(big.NewInt(usage.InputTokens), big.NewInt(usage.CacheReadTokens))
	prompt.Add(prompt, big.NewInt(usage.CacheWriteTokens))
	rates := price.Rates
	if price.Context == ContextTiered && (!prompt.IsInt64() || prompt.Int64() >= price.LongContextFrom) {
		rates = price.LongContext
	}
	numerator := new(big.Int)
	for _, part := range [][2]int64{{usage.InputTokens, rates.InputPerMTok}, {usage.OutputTokens, rates.OutputPerMTok}, {usage.CacheReadTokens, rates.CacheReadPerMTok}, {usage.CacheWriteTokens, rates.CacheWritePerMTok}} {
		if part[1] < 0 {
			return 0, errors.New("negative model price")
		}
		numerator.Add(numerator, new(big.Int).Mul(big.NewInt(part[0]), big.NewInt(part[1])))
	}
	numerator.Mul(numerator, big.NewInt(1000))
	quotient, remainder := new(big.Int), new(big.Int)
	quotient.QuoRem(numerator, big.NewInt(1_000_000), remainder)
	if remainder.Sign() > 0 {
		quotient.Add(quotient, big.NewInt(1))
	}
	quotient.Add(quotient, new(big.Int).Mul(big.NewInt(price.FlatPerCall), big.NewInt(1000)))
	if !quotient.IsInt64() {
		return 0, errors.New("model cost overflow")
	}
	return quotient.Int64(), nil
}
